internal/driver/driver.go

```diff
@@ -0,0 +1,147 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+//
+// Copyright (C) 2018 Canonical Ltd
+// Copyright (C) 2018-2019 IOTech Ltd
+//
+// SPDX-License-Identifier: Apache-2.0
+
+// Package driver this package provides a simple example implementation of
+// ProtocolDriver interface.
+//
+package driver
+
+import (
+	"fmt"
+	"time"
+
+	dsModels "github.com/edgexfoundry/device-sdk-go/pkg/models"
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/logger"
+	contract "github.com/edgexfoundry/go-mod-core-contracts/models"
+)
+
+type Driver struct {
+	lc         logger.LoggingClient
+	asyncCh    chan<- *dsModels.AsyncValues
+	gpiodevice *GPIODev
+}
+
+// Initialize performs protocol-specific initialization for the device
+// service.
+func (s *Driver) Initialize(lc logger.LoggingClient, asyncCh chan<- *dsModels.AsyncValues, deviceCh chan<- []dsModels.DiscoveredDevice) error {
+	s.lc = lc
+	s.asyncCh = asyncCh
+
+	s.gpiodevice = NewGPIODev(lc)
+	return nil
+}
+
+// HandleReadCommands triggers a protocol Read operation for the specified device.
+func (s *Driver) HandleReadCommands(deviceName string, protocols map[string]contract.ProtocolProperties, reqs []dsModels.CommandRequest) (res []*dsModels.CommandValue, err error) {
+
+	s.lc.Info(fmt.Sprintf("protocols: %v resource: %v attributes: %v", protocols, reqs[0].DeviceResourceName, reqs[0].Attributes))
+	gpioNumStr := reqs[0].Attributes["number"]
+
+	now := time.Now().UnixNano()
+
+	for _, req := range reqs {
+		switch req.DeviceResourceName {
+		case "Power":
+			{
+				value, err := s.gpiodevice.GetGPIO(gpioNumStr)
+				if err != nil {
+					return nil, err
+				}
+				cv, _ := dsModels.NewBoolValue(req.DeviceResourceName, now, value)
+				res = append(res, cv)
+			}
+		case "Switch":
+			{
+				value, err := s.gpiodevice.GetGPIO(gpioNumStr)
+				if err != nil {
+					return nil, err
+				}
+				cv, _ := dsModels.NewBoolValue(req.DeviceResourceName, now, value)
+				res = append(res, cv)
+			}
+		}
+	}
+
+	return res, nil
+}
+
+// HandleWriteCommands passes a slice of CommandRequest struct each representing
+// a ResourceOperation for a specific device resource.
+// Since the commands are actuation commands, params provide parameters for the individual
+// command.
+func (s *Driver) HandleWriteCommands(deviceName string, protocols map[string]contract.ProtocolProperties, reqs []dsModels.CommandRequest,
+	params []*dsModels.CommandValue) error {
+	s.lc.Info(fmt.Sprintf("Driver.HandleWriteCommands: protocols: %v, resource: %v, attribute: %v, parameters: %v", protocols, reqs[0].DeviceResourceName, reqs[0].Attributes, params))
+	gpioNumStr := reqs[0].Attributes["number"]
+
+	for _, param := range params {
+		var err error
+		switch param.DeviceResourceName {
+		case "Power":
```

Comment:
Please see my previous statement about use of a switch statement to check device resource names.

I think the intention might be to ensure handling of read/write, as the Read switch statement above has cases for Power and Switch (both readable), and this switch has cases for Power and Led (both writable). The Go SDK already handles this for you (i.e. if a write command is received for a device resource that is read-only, an error is returned).